src/dev/so_migration/compare_snapshots.ts

```diff
@@ -0,0 +1,164 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { ToolingLog } from '@kbn/tooling-log';
+import { readFile } from 'fs/promises';
+import { existsSync } from 'fs';
+import * as os from 'os';
+import { execSync } from 'child_process';
+import { basename, resolve } from 'path';
+import { MigrationInfoRecord, MigrationSnapshot } from './types';
+
+const SO_MIGRATIONS_SNAPSHOT_FOLDER = 'kibana-so-types-snapshots';
+
+interface CompareSnapshotsParameters {
+  from: string;
+  to: string;
+  log: ToolingLog;
+  outputPath?: string;
+}
+
+async function compareSnapshots({
+  outputPath,
+  log,
+  from,
+  to,
+}: CompareSnapshotsParameters): Promise<any> {
+  validateInput({
+    from,
+    to,
+  });
+
+  const fromSnapshotPath = isCommitHash(from) ? await downloadSnapshot(from, log) : from;
+  const toSnapshotPath = isCommitHash(to) ? await downloadSnapshot(to, log) : to;
+
+  const fromSnapshot = await loadJson(fromSnapshotPath);
+  const toSnapshot = await loadJson(toSnapshotPath);
+
+  const result = compareSnapshotFiles(fromSnapshot, toSnapshot);
+
+  if (outputPath) {
+    log.info(`Snapshots compared: ${from} <=> ${to}`);
+    writeSnapshot(outputPath, result);
+    log.info(`Output written to: ${outputPath}`);
+  } else {
+    log.info(`Snapshots compared: ${from} <=> ${to}`);
+    log.info(result);
+  }
+
+  return result;
+}
+
+function validateInput({ from, to }: { from: string; to: string }) {
+  if (!from || !to) {
+    throw new Error('"--from" and "--to" must be specified');
+  }
+
+  if (from === to) {
+    throw new Error('"from" and "to" must be different');
+  }
+
+  if (!isCommitHash(from) && !isUrl(from) && !isFilePath(from)) {
+    throw new Error(
+      'Invalid "--from" argument. Must be a full commit hash, a URL or an existent file path'
+    );
+  }
+
+  if (!isCommitHash(to) && !isUrl(to) && !isFilePath(to)) {
+    throw new Error(
+      'Invalid "--to" argument. Must be a full commit hash, a URL or an existent file path'
+    );
+  }
+}
+
+function writeSnapshot(outputPath: string, result: any) {
+  throw new Error('Function not implemented.');
+}
+
+function isCommitHash(str: string) {
+  return /^[a-f0-9]{40}$/.test(str);
+}
+
+function isUrl(str: string) {
+  return /^https?:\/\//.test(str);
+}
+
+function isFilePath(str: string) {
+  return existsSync(str);
+}
+
+async function downloadToTemp(googleCloudUrl: string): Promise<string> {
+  const fileName = basename(googleCloudUrl);
+  const filePath = resolve(os.tmpdir(), fileName);
+
+  execSync(`gsutil cp ${googleCloudUrl} ${filePath}`);
+
+  return filePath;
+}
+
+async function downloadSnapshot(commitHash: string, log: ToolingLog): Promise<string> {
+  const googleCloudUrl = `gs://${SO_MIGRATIONS_SNAPSHOT_FOLDER}/${commitHash}.json`;
```

Comment:
Do we expect this to be used locally with any frequency?  I think the answers no, but if not it may be helpful to download over http so devs don't have to install gsutil.

Reply:
Even if running local is not the main goal, if we can make it easier without downsides, we probably should.

Reply:
I think I ultimately settled for the `gsutil` way, because there, the authentication can be ensured. 

However, if this is not sensitive or data that can be abused, then we can make the snapshots public, and we can do it over http. @pgayvallet - this is probably OK to be made public, there's nothing in it that's not public through the github repo anyway, right?

Reply:
Yeah, I don't think it can be considered sensitive information, everything can be re generated from the repo anyway, so having it read-public seems fine

Reply:
added in: [da9738d](https://github.com/elastic/kibana/pull/168623/commits/da9738d6dd7f82c9e8200ce1a60495782febf6fa)